lib/dc_threaded_queue.c

```diff
@@ -0,0 +1,177 @@
+/**
+* karel on 2021-04-10.
+*/
+
+#include <errno.h>
+#include <stdlib.h>
+#include "dc_threaded_queue.h"
+#include "stdio.h"
+#include "fcntl.h"
+
+#define DC_THREADED_QUEUE_SEM_NAME "/dc_threaded_queue_%zu_sem"
+
+
+static size_t queue_counter = 0;
+
+void get_sem_name( char buff[255], size_t val ) {
+    sprintf( buff, DC_THREADED_QUEUE_SEM_NAME, val );
+}
+
+sem_t * open_sem( size_t value, size_t * id ) {
+    sem_t * new_sem = NULL;
+
+    size_t initial_counter = queue_counter;
+
+    while ( !new_sem ) {
+        char name[255] = { 0 };
+        get_sem_name( name, queue_counter );
+
+        new_sem = sem_open( name, O_CREAT | O_EXCL, 0640, value );
+        *id = queue_counter;
+
+        if ( new_sem == SEM_FAILED) {
+            if ( errno == EINVAL || errno == ENOMEM || errno == EMFILE ) {
+                return NULL;
+            }
+        }
+
+        queue_counter++; // overflow is acceptable
+
+        if ( queue_counter == initial_counter ) {
+            // overflowed and came back around to the initial position
+            // practical limit for max number of queues with the current name implementation
+            return NULL;
+        }
+    }
+
+    return new_sem;
+}
+
+void close_sem( size_t sem_id ) {
+    char name[255] = { 0 };
+    get_sem_name( name, sem_id );
+    sem_unlink( name );
+}
+
+
+int add( struct dc_threaded_queue * queue, void * element );
+
+void * take( struct dc_threaded_queue * queue );
```

Comment:
Should these go into the header?

Reply:
I am encapsulating the behaviour as part of the struct in this datastructure. Experimenting on how to control namespace pollution.